docs/user-guide/performance.rst

```diff
@@ -0,0 +1,332 @@
+and ``z[40:60]``) then each worker will be writing to a separate chunk and no
+synchronization is required. However, if two workers are running and each
+attempts to write to a 30 element region (i.e., ``z[0:30]`` and ``z[30:60]``)
+then it is possible both workers will attempt to modify the middle chunk at the
+same time, and synchronization is required to prevent data loss.
+
+Zarr provides support for chunk-level synchronization. E.g., create an array
+with thread synchronization:
+
+.. .. ipython:: python
+..    :verbatim:
+
+..    z = zarr.zeros((10000, 10000), chunks=(1000, 1000), dtype='i4', synchronizer=zarr.ThreadSynchronizer())
+..    z
+
+This array is safe to read or write within a multi-threaded program.
+
+Zarr also provides support for process synchronization via file locking,
+provided that all processes have access to a shared file system, and provided
+that the underlying file system supports file locking (which is not the case for
+some networked file systems). E.g.:
+
+.. .. ipython:: python
+..    :verbatim:
+
+..    synchronizer = zarr.ProcessSynchronizer('data/example.sync')
+
+..    z = zarr.open_array('data/example', mode='w', shape=(10000, 10000), chunks=(1000, 1000), dtype='i4', synchronizer=synchronizer)
+..    z
+
+This array is safe to read or write from multiple processes.
+
+When using multiple processes to parallelize reads or writes on arrays using the Blosc
+compression library, it may be necessary to set ``numcodecs.blosc.use_threads = False``,
+as otherwise Blosc may share incorrect global state amongst processes causing programs
+to hang. See also the section on :ref:`tutorial_tips_blosc` below.
+
+Please note that support for parallel computing is an area of ongoing research
+and development. If you are using Zarr for parallel computing, we welcome
+feedback, experience, discussion, ideas and advice, particularly about issues
+related to data integrity and performance.
+
+.. _tutorial_pickle:
+
+Pickle support
+--------------
+
+Zarr arrays and groups can be pickled, as long as the underlying store object can be
+pickled. With the exception of the :class:`zarr.storage.MemoryStore`, any of the
+storage classes provided in the :mod:`zarr.storage` module can be pickled.
+
+If an array or group is backed by a persistent store such as the a :class:`zarr.storage.LocalStore`,
+:class:`zarr.storage.ZipStore` or :class:`zarr.storage.FsspecStore` then the store data
+**are not** pickled. The only thing that is pickled is the necessary parameters to allow the store
+to re-open any underlying files or databases upon being unpickled.
+
+E.g., pickle/unpickle an local store array:
+
+.. ipython:: python
+
+   import pickle
+
+   z1 = zarr.array(store="data/example-2", data=np.arange(100000))
+   s = pickle.dumps(z1)
+   z2 = pickle.loads(s)
+   z1 == z2
+   np.all(z1[:] == z2[:])
+
+.. _tutorial_tips_blosc:
+
+Configuring Blosc
+~~~~~~~~~~~~~~~~~
```

Comment:
The heading level is wrong here, I don't think this is meant to be under "Pickle support"